### PR TITLE
Guanyu/rewrite/scenariotest

### PR DESCRIPTION
**Pull request by JBGYZ**
1. Finished scenario test with http
2. added 2 fields in getelectioninfo response

### CODE REVIEW

File: integration/scenario_test.go [hunk 1]
```diff
@@ -0,0 +1,830 @@
...
+
+		t.Parallel()
+		proxyAddr1 := "http://localhost:8081"
+		proxyAddr2 := "http://localhost:8082"
+		proxyAddr3 := "http://localhost:8083"
+
+		proxyArray := [3]string{proxyAddr1, proxyAddr2, proxyAddr3}
+
+		// ###################################### CREATE SIMPLE ELECTION ######
+		create_election_js := `{"Configuration":{"MainTitle":"electionTitle","Scaffold":[{"ID":"YWE=","Title":"subject1","Order":null,"Subjects":null,"Selects":[{"ID":"YmI=","Title":"Select your favorite snacks","MaxN":3,"MinN":0,"Choices":["snickers","mars","vodka","babibel"]}],"Ranks":[],"Texts":null},{"ID":"ZGQ=","Title":"subject2","Order":null,"Subjects":null,"Selects":null,"Ranks":null,"Texts":[{"ID":"ZWU=","Title":"dissertation","MaxN":1,"MinN":1,"MaxLength":3,"Regex":"","Choices":["write yes in your language"]}]}]},"AdminID":"adminId"}`
```
Comment (emduc) [inline integration/scenario_test.go:63]: would be good to add some returns so that we can read the content of the configuration easily

**Review by emduc — COMMENTED** [review]
I made a quick pass on it since I am starting to try and solve this shuffle problem, this new version will definitely be helpful thank you :)

I know it wasn't the case in the original scenario test but it could be nice to have the test be generic in the number of nodes  and votes, similarly to what is done in integration_test.go. It should not be too difficult to implement (just a matter of having for eg arrays of addresses instead of addr1 addr2 and addr3) but might take some time.
edit:  actually that might be more annoying since we need to launch the nodes seperately

Depending on who merges first, the problem of signed request will also arise since the proxy will soon require the requests to be signed.

File: integration/scenario_test.go [hunk 2]
```diff
@@ -0,0 +1,1097 @@
...
+	}
+}
+
+func testVariableNode(numNodes int, numVotes int, numElection int) func(*testing.T) {
+	return func(t *testing.T) {
+
+		proxyList := make([]string, numNodes)
+
+		for i := 0; i < numNodes; i++ {
+			if i < 10 {
```
Comment (nkcr) [inline integration/scenario_test.go:389]: very hacky. What you want to do is just:
```go
 proxyList[i] = fmt.Sprintf("http://localhost:90%02d", 80+i)
```

File: integration/scenario_test.go [hunk 3]
```diff
@@ -0,0 +1,1097 @@
...
+			} else {
+				proxyList[i] = "http://localhost:909" + strconv.Itoa(i-10)
+			}
+			fmt.Println(proxyList[i])
+		}
+
+		var wg sync.WaitGroup
+
+		for i := 0; i < numElection; i++ {
+			fmt.Println("Starting worker", i)
```
Comment (nkcr) [inline integration/scenario_test.go:400]: `t.log` instead
(same for other places)

File: integration/scenario_test.go [hunk 4]
```diff
@@ -0,0 +1,1097 @@
...
+	secretkeyBuf, err := hex.DecodeString("28912721dfd507e198b31602fb67824856eb5a674c021d49fdccbe52f0234409")
+	require.NoError(t, err, "failed to decode key: %v", err)
+
+	secret := suite.Scalar()
+	err = secret.UnmarshalBinary(secretkeyBuf)
+	require.NoError(t, err, "failed to Unmarshal key: %v", err)
+
+	// ###################################### CREATE SIMPLE ELECTION ######
+
+	t.Logf("Create election")
```
Comment (nkcr) [inline integration/scenario_test.go:427]:
```suggestion
	t.Log("Create election")
```

File: integration/scenario_test.go [hunk 5]
```diff
@@ -0,0 +1,1097 @@
...
+	}
+	require.True(t, tmpCount, "front end votes are different from decrypted votes")
+
+}
+
+// -----------------------------------------------------------------------------
+// Utility functions
+func marshallBallotManual(voteStr string, pubkey kyber.Point, chunks int) (ptypes.CiphervoteJSON, error) {
+
+	var ballot = make(ptypes.CiphervoteJSON, chunks)
```
Comment (nkcr) [inline integration/scenario_test.go:703]:
```suggestion
	ballots := make(ptypes.CiphervoteJSON, chunks)
```

File: integration/scenario_test.go [hunk 6]
```diff
@@ -0,0 +1,1097 @@
...
+		ballot[i] = ptypes.EGPairJSON{
+			K: kbuff,
+			C: cbuff,
+		}
+	}
+
+	return ballot, nil
+}
+
+func EncryptManual(message []byte, pubkey kyber.Point) (K, C kyber.Point, remainder []byte, err error) {
```
Comment (nkcr) [inline integration/scenario_test.go:743]:
```suggestion
func encryptManual(message []byte, pubkey kyber.Point) (K, C kyber.Point, remainder []byte, err error) {
```

File: integration/scenario_test.go [hunk 7]
```diff
@@ -0,0 +1,1097 @@
...
+	return K, C, remainder, nil
+}
+
+func ChunksPerBallotManuel(BallotSize int) int {
+	if BallotSize%29 == 0 {
+		return BallotSize / 29
+	}
+
+	return BallotSize/29 + 1
+}
```
Comment (nkcr) [inline integration/scenario_test.go:767]:
```suggestion
func chunksPerBallot(size int) int { return (size-1) / 29 + 1}
```

File: integration/scenario_test.go [hunk 8]
```diff
@@ -0,0 +1,1097 @@
...
+
+// Encode implements serde.FormatEngine
+func EncodeCiphervote(ciphervote types.Ciphervote) ([]byte, error) {
```
Comment (nkcr) [inline integration/scenario_test.go:770]:
```suggestion
func encodeCiphervote(ciphervote types.Ciphervote) ([]byte, error) {
```

File: integration/scenario_test.go [hunk 9]
```diff
@@ -0,0 +1,1097 @@
...
+	data, err := json.Marshal(m)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to marshal cipher vote json: %v", err)
+	}
+
+	return data, nil
+}
+
+// CiphervoteJSON is the JSON representation of a ciphervote
+type CiphervoteJSON []EGPairJSON
```
Comment (nkcr) [inline integration/scenario_test.go:800]: you should take the definition from `proxy/types`

File: integration/scenario_test.go [hunk 10]
```diff
@@ -0,0 +1,1097 @@
...
+
+// EGPairJSON is the JSON representation of an ElGamal pair
+type EGPairJSON struct {
+	K []byte
+	C []byte
+}
+
+// Unmarshal decodes the given string according to the format described in
+// "state of smart contract.md"
+func UnmarshalBallotManual(marshalledBallot string, configuration types.Configuration) (types.Ballot, error) {
```
Comment (nkcr) [inline integration/scenario_test.go:810]:
```suggestion
func unmarshalBallot(ballot string, conf types.Configuration) (types.Ballot, error) {
```

Comment (nkcr) [reply integration/scenario_test.go:810]: Please use the function already defined in `types/ballots.go`.

File: runNode.sh [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+from=0
+fromb=1
+to=$1
+ne=0
+while [ $fromb -le $to ]
+do
+if [ $from -le 9 ]
+then
+gnome-terminal --tab --title="test" --command="./memcoin --config /tmp/node$fromb start --postinstall --promaddr :910$from --proxyaddr :908$from --proxykey $pk --listen tcp://0.0.0.0:200$fromb --public //localhost:200$fromb"
```
Comment (nkcr) [inline runNode.sh:17]: this won't work on a standard UNIX environment (`gnome-terminal`).

File: integration/scenario_test.go [hunk 12]
```diff
@@ -0,0 +1,1097 @@
...
+	}
+
+	signedJSON, err := json.Marshal(signed)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to create json signed: %v", err)
+	}
+
+	return signedJSON, nil
+}
+func initDKG(secret kyber.Scalar, proxyAddr, electionIDHex string, t *testing.T) error {
```
Comment (nkcr) [inline integration/scenario_test.go:1031]: missing newline above

File: runNode.sh (hunk 11 above)
Comment (nkcr) [reply runNode.sh:17]: (and it should)

Comment (nkcr) [reply runNode.sh:17]: (and it should)

**Review by nkcr — APPROVED** [review]
LGTM 🚀

**Pull request merged**